File: 0002-add-two-numbers/0002-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode dummy = new ListNode(0);
        ListNode sumCur = dummy;
        ListNode l1Cur = l1;
        ListNode l2Cur = l2;

        boolean prev = false;
        while (l1Cur != null || l2Cur != null) {
            int cur = prev ? 1 : 0;
            prev = false;

            if (l1Cur != null) {
                cur += l1Cur.val;
                l1Cur = l1Cur.next;
            }

            if (l2Cur != null) {
                cur += l2Cur.val;
                l2Cur = l2Cur.next;
            }

            if (cur >= 10) {
                prev = true;
                cur %= 10;
            }
            ListNode node = new ListNode(cur);
            sumCur.next = node;
            sumCur = sumCur.next;
        }

        if (prev) {
            ListNode node = new ListNode(1);
            sumCur.next = node;
            sumCur = sumCur.next;            
        }

        return dummy.next;
    }
}

File: 0002-add-two-numbers/0002-add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        ListNode* cur1 = l1;
        ListNode* cur2 = l2;
        ListNode dummy(0);
        ListNode* curSum = &dummy;

        bool tenFlag = false;
        while (cur1 != nullptr || cur2 != nullptr) {
            int curVal = tenFlag ? 1 : 0;
            tenFlag = false;

            if (cur1 != nullptr) {
                curVal += cur1->val;
                cur1 = cur1->next;
            }

            if (cur2 != nullptr) {
                curVal += cur2->val;
                cur2 = cur2->next;
            }    

            ListNode* curNode = new ListNode(curVal % 10);
            tenFlag = (curVal / 10 == 1);
            curSum->next = curNode;
            curSum = curSum->next;        
        }

        if (tenFlag) {
            ListNode* curNode = new ListNode(1);
            curSum->next = curNode;
            curSum = curSum->next;              
        }

        return dummy.next;
    }
};